cranelift/wasm/src/code_translator.rs

```diff
@@ -3579,50 +3581,260 @@ where
+    // Runtime code that is generated:
+    //
+    // (compile-time) num_required_st2g_loops = floor(size / LOOP_UNROLL_BYTES_THRESHHOLD)
+    // (compile-time) num_required_last_stg = (size % LOOP_UNROLL_BYTES_THRESHHOLD) / 16
+    //
+    // (compile-time) if num_required_st2g_loops <= 1
+    // (compile-time)     generate WITHOUT_LOOP_UNROLLING_THRESHHOLD
+    // (compile-time) else
+    // (compile-time)     generate WITH_LOOP_UNROLLING_THRESHHOLD
+    //
+    // WITHOUT_LOOP_UNROLLING_THRESHHOLD:
+    // +---less_than_threshhold
+    // | st2g(tagged_ptr, iter_ptr + 0)          \
+    // | st2g(tagged_ptr, iter_ptr + 32)          | repeat `n` times, i.e. while (remaining_bits >= 32) holds
+    // | ...                                      |
+    // | st2g(tagged_ptr, iter_ptr + (n-1) * 32) /
+    // |
+    // | stg(tagged_ptr, iter_ptr + n * 32)  | generate only if required
+    // +---
+    //
+    // WITH_LOOP_UNROLLING_THRESHHOLD:
+    // st2g_loop_counter = 0
+    // +---st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+    // | if (st2g_loop_counter < num_required_full_st2g_loops)
+    // |     goto st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+    // | else
+    // |     (compile-time) if (num_required_last_stgs == 0)
+    // |                        goto next_block
+    // |     (compile-time) else
+    // |                        goto last_stgs_block
+    // |
+    // | +---st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+    // | | st2g(tagged_ptr, iter_ptr + 0)          \
+    // | | st2g(tagged_ptr, iter_ptr + 32)          | repeat `n`=LOOP_UNROLL_BYTES_THRESHHOLD times
+    // | | ...                                      |
+    // | | st2g(tagged_ptr, iter_ptr + (n-1) * 32) /
+    // | |
+    // | | iter_ptr += LOOP_UNROLL_BYTES_THRESHHOLD
+    // | | st2g_loop_counter++
+    // | | goto st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+    // | +---
+    // +---
+    //
+    // +---last_stgs_block
+    // | (compile-time) offset = num_required_st2g_loops * LOOP_UNROLL_BYTES_THRESHHOLD
+    // |
+    // | st2g(tagged_ptr, iter_ptr + offset + 0)          \
+    // | st2g(tagged_ptr, iter_ptr + offset + 32)          | repeat `n` times, i.e. while (remaining_bits >= 32) holds
+    // | ...                                               |
+    // | st2g(tagged_ptr, iter_ptr + offset + (n-1) * 32) /
+    // |
+    // | stg(tagged_ptr, iter_ptr + offset + n * 32)  | generate only if required
+    // +---
+    //
+    // +---next_block
+    // +---
+
+    // TODO: only use one of i64 or u64
+
+    // If the number of bytes to be tagged is larger than this threshhold,
+    // insert a loop that tags using `st2g` chunks.
+    // Threshhold should be a multiple of 32, so each `st2g` chunk can be
+    // exactly divided into only 32-byte tagging `st2g` instructions
+    // This constant was adapted from clang.
+    const LOOP_UNROLL_BYTES_THRESHHOLD_I64: i64 = 160;
+    const LOOP_UNROLL_BYTES_THRESHHOLD_U64: u64 = 160;
```

Comment:
Any ideas how to make this a little nicer by not having to define the threshhold constant twice? The problem is that it is mostly required as an i64, but a few lines below we have to use it in a division, which would be better suited as a u64 division.